lxd/storage/dir.go

```diff
@@ -0,0 +1,532 @@
+	return nil
+}
+
+// VolumeCopy copies the storage volume including the provided snapshots.
+func (s *dir) VolumeCopy(project string, source string, target string, snapshots []string, volumeType VolumeType) error {
+	var sourceMountPoint string
+	var targetMountPoint string
+
+	switch volumeType {
+	case VolumeTypeContainer:
+		for _, snap := range snapshots {
+			sourceMountPoint = GetSnapshotMountPoint(project, s.pool.Name, fmt.Sprintf("%s/%s", source, snap))
+			targetMountPoint = GetSnapshotMountPoint(project, s.pool.Name, fmt.Sprintf("%s/%s", target, snap))
+
+			err := s.rsync(sourceMountPoint, targetMountPoint)
+			if err != nil {
+				return err
+			}
+		}
+
+		sourceMountPoint = GetContainerMountPoint(project, s.pool.Name, source)
+		targetMountPoint = GetContainerMountPoint(project, s.pool.Name, target)
+	case VolumeTypeCustom:
+		for _, snap := range snapshots {
+			sourceMountPoint = GetStoragePoolVolumeSnapshotMountPoint(s.pool.Name, fmt.Sprintf("%s/%s", source, snap))
+			targetMountPoint = GetStoragePoolVolumeSnapshotMountPoint(s.pool.Name, fmt.Sprintf("%s/%s", target, snap))
+
+			err := s.rsync(sourceMountPoint, targetMountPoint)
+			if err != nil {
+				return err
+			}
+		}
+
+		sourceMountPoint = GetStoragePoolVolumeMountPoint(s.pool.Name, source)
+		targetMountPoint = GetStoragePoolVolumeMountPoint(s.pool.Name, target)
+	default:
+		return fmt.Errorf("Unsupported volume type: %v", volumeType)
+	}
+
+	return s.rsync(sourceMountPoint, targetMountPoint)
+}
+
+// VolumeDelete removes the given storage volume.
+func (s *dir) VolumeDelete(project string, volumeName string, recursive bool, volumeType VolumeType) error {
+	var path string
+
+	switch volumeType {
+	case VolumeTypeContainer:
+		path = GetContainerMountPoint(project, s.pool.Name, volumeName)
+	case VolumeTypeCustom:
+		path = GetStoragePoolVolumeMountPoint(s.pool.Name, volumeName)
+	default:
+		return fmt.Errorf("Unsupported volume type: %v", volumeType)
+	}
+
+	err := os.RemoveAll(path)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// VolumeRename renames the given storage volume.
+func (s *dir) VolumeRename(project string, oldName string, newName string, snapshots []string,
+	volumeType VolumeType) error {
+	var oldPath string
+	var newPath string
+
+	switch volumeType {
+	case VolumeTypeContainer:
+		oldPath = GetContainerMountPoint(project, s.pool.Name, oldName)
+		newPath = GetContainerMountPoint(project, s.pool.Name, newName)
+	case VolumeTypeCustom:
+		oldPath = GetStoragePoolVolumeMountPoint(s.pool.Name, oldName)
+		newPath = GetStoragePoolVolumeMountPoint(s.pool.Name, newName)
+	default:
+		return fmt.Errorf("Unsupported volume type: %v", volumeType)
+	}
+
+	if shared.PathExists(newPath) {
+		// Nothing to do
+		return nil
+	}
+
+	return os.Rename(oldPath, newPath)
+}
+
+// VolumeMount mounts the given storage volume.
+func (s *dir) VolumeMount(project string, name string, volumeType VolumeType) (bool, error) {
+	// Nothing to do
+	return true, nil
```

Comment:
@monstermunchkin why does this function return both a bool and a err? Can it be modified such that a nil-err indicates success? (I may be missing something here :))

Reply:
The bool indicates whether the function actually mounted to volume. If the function returns `false`, it doesn't automatically mean there was an error. I don't think we should change the return values.

Reply:
@monstermunchkin ah, I see, so it could already be mounted? Do you know what scenario this would happen in?

Reply:
Yes, it could already be mounted but I don't know what scenario this would happen in. I guess it's called sometimes just to makes sure something's really mounted.